A0-950 - The button game [implementation of the scenario 2] (#470)
Cardinal-Cryptography/aleph-node: contracts/README.md

# TheButton

Series of smart contract based games, with varying tokenomics.
TheButton is based on the famous [game](https://en.wikipedia.org/wiki/The_Button_(Reddit)) on reddit.

> - Button lives for a set time.

Reviewer: ```suggestion
- TheButton lives for a set time.
```

> - Pressing the button extends it's life.

Reviewer: ```suggestion
- Pressing the button extends its life.
```

- Users are rewarded for playing the game.
- Everybody can play only once.

```
  |______|_______|
  |      |       |
start   now    deadline
```

## Red Button

There is a pre-minted amount of red tokens (a classic ERC20).

> Users are rewarded for clicking as early on as possible, maximizing TheButtons life.

Reviewer: ```suggestion
Users are rewarded for clicking as early on as possible, maximizing TheButton's life.
```

> Users are rewarded for clicking as early on as possible, maximizing TheButton's life.

Author: reworded

```
score = deadline - now
```

There are two built-in incentives:

> * playing for score: If you clicked in 10'th second of TheButtons life set for example to 900 blocks you get rewarded based on score of 900-10=890 (and the button's life now will end at block 910).

Reviewer: ```suggestion
* playing for score: If you clicked in the 10th second of TheButton's life set for example to 900 blocks you get rewarded based on score of 900-10=890 (and the button's life now will end at block 910).
```

> * playing to be ThePressiah: the last player to click get's 50% of the total rewards pool.

Reviewer: ```suggestion
* playing to be ThePressiah: the last player to click gets 50% of the total reward pool.
```

## Yellow button

> Similar to button red, but in that scenario the awards are reversed - players get rewarded for extending the button's life further into the future, i.e.:

Reviewer: ```suggestion
Similar to Red Button, but in this scenario the rewards are reversed - players get rewarded for extending the button's life further into the future, i.e.:
```

```
score = now - start
```

## Blue button

Game continues in perpetuity (but in practice as long as there are accounts that can still play it)
- In each iteration of the game TheButton lives for a number of blocks

> - Clicking TheButton resets it's countdown timer (extending the button's life further into the future)

Reviewer: ```suggestion
- Clicking TheButton resets its countdown timer (extending the button's life further into the future)
```

> - Blue Tokens are continuously minted at the end of each iteration

Reviewer: I assume the yellow button generates yellow tokens?

```suggestion
- Yellow Tokens are continuously minted at the end of each iteration
```

Author: this is a description of another scenarios, there will be three games in total

- Players are rewarded for playing, with the ultimate goal of being the Pressiah (the last person to click the button)
- Reward rules:

>   - If you’re not ThePressiah, you get *k* tokens if you pressed the button as *k-th* person in a row.

Reviewer: ```suggestion
  - If you’re not ThePressiah, you get *k* tokens if you pressed the button as the *k-th* person in a row.
```

  - ThePressiah of the iteration with *k* button presses gets [k*(k+1)]/2 tokens.

# Development

Firstly bootstrap a one-node chain based on our `smartnet` chain:

```bash
 ./.github/scripts/run_smartnet.sh
```

Secondly `deploy` script takes care of compiling, deploying and interacting with the contract(s):

```bash
./contracts/scripts/deploy.sh
```

Specifically it will:
- deploy the ERC20 token contract
- deploy the game contract
- transfer token balance to the game contract
- whitelist some accounts for playing the game
- interact with the game from the whitelisted account

> - wait past the game deadline, trigger game end and rewards distribution

Reviewer: ```suggestion
- wait past the game deadline, trigger game end and reward distribution
```
